### PR TITLE
feat(worker/activitypub): modify mastodon data transformation logic

### PR DESCRIPTION
**Pull request by FrankLi123**
## Summary

This PR modifies the Mastodon worker transformation logic to addresses identified issues.

Note: There are 2 `ToDo` comments to be checked.

**`Create` Note Object:**
```
{
  "id": "https://digitalcourage.social/users/Volksverpetzer/statuses/113287832117292671/activity",
  "network": "mastodon",
  "index": 0,
  "from": "@Volksverpetzer@digitalcourage.social",
  "to": "@Volksverpetzer@digitalcourage.social",
  "tag": "social",
  "type": "post",
  "platform": "Mastodon",
  "total_actions": 1,
  "actions": [
    {
      "tag": "social",
      "type": "post",
      "platform": "Mastodon",
      "from": "@Volksverpetzer@digitalcourage.social",
      "to": "@Volksverpetzer@digitalcourage.social",
      "metadata": {
        "handle": "@Volksverpetzer@digitalcourage.social",
        "body": "Die Lügen der Rechten kosten uns Millionen: Arbeitende Schutzsuchende schieben wir ab und die steigenden Kosten müssen junge Menschen zahlen. Der Rassismus von AfD & Co. wird junge Menschen teuer zu stehen kommen. https://www. volksverpetzer.de/analyse/woll en-nicht-fuer-rassismus-zahlen/?utm_source=mstdn ( https://www.volksverpetzer.de/analyse/wollen-nicht-fuer-rassismus-zahlen/?utm_source=mstdn )",
        "publication_id": "113287832117292671",
        "timestamp": 1728635133
      },
      "related_urls": [
        "https://digitalcourage.social/users/Volksverpetzer/statuses/113287832117292671"
      ]
    }
  ],
  "success": true,
  "timestamp": 1728635133
}
```

**`Create` Note Object with Media:**
```
 {
          "id": "https://mastodon.world/users/ctxt/statuses/113287758853959995/activity",
          "network": "mastodon",
          "index": 0,
          "from": "@ctxt@mastodon.world",
          "to": "@ctxt@mastodon.world",
          "tag": "social",
          "type": "post",
          "platform": "Mastodon",
          "total_actions": 1,
          "actions": [
            {
              "tag": "social",
              "type": "post",
              "platform": "Mastodon",
              "from": "@ctxt@mastodon.world",
              "to": "@ctxt@mastodon.world",
              "metadata": {
                "handle": "@ctxt@mastodon.world",
                "body": "Y por si se les pasó, les dejamos también la pieza que nos mandó Lola Matamala contándonos en primera persona el desahucio que sufrió su madre el mes pasado después de llevar setenta años viviendo en la misma calle.\n\nLa empresa Dapamali Works compró el edifio y fue echando uno por uno a todos los vecinos:\n\nhttps:// ctxt.es/es/20241001/Firmas/476 02/lola-matamala-desahucio-Dapamali-Works-primera-persona-aurora-getafe.htm ( https://ctxt.es/es/20241001/Firmas/47602/lola-matamala-desahucio-Dapamali-Works-primera-persona-aurora-getafe.htm )",
                "media": [
                  {
                    "address": "https://s3.eu-central-2.wasabisys.com/mastodonworld/media_attachments/files/113/287/758/697/714/115/original/2c6c341deae150c0.jpg",
                    "mime_type": "image/jpeg"
                  }
                ],
                "publication_id": "113287758853959995",
                "timestamp": 1728634015
              },
              "related_urls": [
                "https://mastodon.world/users/ctxt/statuses/113287758853959995"
              ]
            }
          ],
          "success": true,
          "timestamp": 1728634015
}
```

**`Comment` Note Object:**
```
{
  "id": "https://social.timespiral.co.jp/users/find575/statuses/113287741759175713/activity",
  "network": "mastodon",
  "index": 0,
  "from": "@find575@social.timespiral.co.jp",
  "to": "@cs133@pawoo.net",
  "tag": "social",
  "type": "comment",
  "platform": "Mastodon",
  "total_actions": 1,
  "actions": [
    {
      "tag": "social",
      "type": "comment",
      "platform": "Mastodon",
      "from": "@find575@social.timespiral.co.jp",
      "to": "@cs133@pawoo.net",
      "metadata": {
        "handle": "@find575@social.timespiral.co.jp",
        "body": "@ cs133 ( https://pawoo.net/@cs133 ) 俳句を発見しました！\n『各駅に エスペラントで 愛称が』",
        "publication_id": "113287741759175713",
        "timestamp": 1728633754,
        "target": {
          "handle": "@cs133@pawoo.net",
          "body": "賢治に毒されて各駅にエスペラントで愛称が付けられている釜石線",
          "publication_id": "113287741711853329",
          "timestamp": 1728633754
        }
      },
      "related_urls": [
        "https://social.timespiral.co.jp/users/find575/statuses/113287741759175713"
      ]
    }
  ],
  "success": true,
  "timestamp": 1728633754
}
```

**`Announce` Note Object:**
```
{
  "id": "https://fosstodon.org/users/bert_hubert/statuses/113287843427757878/activity",
  "network": "mastodon",
  "index": 0,
  "from": "@bert_hubert@fosstodon.org",
  "to": "@ravirockks@infosec.exchange",
  "tag": "social",
  "type": "share",
  "platform": "Mastodon",
  "total_actions": 1,
  "actions": [
    {
      "tag": "social",
      "type": "share",
      "platform": "Mastodon",
      "from": "@bert_hubert@fosstodon.org",
      "to": "@ravirockks@infosec.exchange",
      "metadata": {
        "handle": "@bert_hubert@fosstodon.org",
        "publication_id": "113287843427757878",
        "timestamp": 1728635306,
        "target": {
          "handle": "@ravirockks@infosec.exchange",
          "body": "THE CYBER RESILIENCE ACT HAS BEEN ADOPTED BY THE COUNCIL OF THE EUROPEAN UNION!\n\nThe EU is weeks away from becoming the first jurisdiction with a bespoke regulatory framework for the product security _and_ labelling of all software sold commercially in the EU (save stuff covered by other EU rules like cars and healthtech).\n\nYes, the Yanks (via EO14028–>NIST) defined critical software (the CRA has ‘important products with digital elements’ and ‘critical products with digital elements’), but the Yanks, for now at least, have only gone down the procurement route for regulating vendor SDLCs. The EU, on the other hand, is covering everything sold commercially (bar the stated exceptions) to anyone in the EU.\n\nBig day for all us SDLC regulation people!\n\nWhat happens next: Council and EuroParl President sign it —> Publication in the EU OJ —> Entry into force 20 days later —> Application of most provisions 36 months later.\n\nPress release (includes link to final text): https://www. consilium.europa.eu/en/press/p ress-releases/2024/10/10/cyber-resilience-act-council-adopts-new-law-on-security-requirements-for-digital-products/ ( https://www.consilium.europa.eu/en/press/press-releases/2024/10/10/cyber-resilience-act-council-adopts-new-law-on-security-requirements-for-digital-products/ )",
          "publication_id": "113286145052908177",
          "timestamp": 1728609391
        }
      },
      "related_urls": [
        "https://fosstodon.org/users/bert_hubert/statuses/113287843427757878"
      ]
    }
  ],
  "success": true,
  "timestamp": 1728635306
}
```

## Checklist

- [x] The commit message follows [Angular Contributing guidelines](https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit);
- [x] Tests for the changes have been added (for bug fixes / features);

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by polebug** [comment]
```
{
          "id": "https://beekeeping.ninja/users/Pagan_Animist/statuses/112840117527501203",
          "network": "mastodon",
          "index": 0,
          "from": "@Pagan_Animist@beekeeping.ninja",
          "to": "@evedazzle@mas.to",
          "tag": "social",
          "type": "comment",
          "platform": "Mastodon",
          "total_actions": 1,
          "actions": [
            {
              "tag": "social",
              "type": "comment",
              "platform": "Mastodon",
              "from": "@Pagan_Animist@beekeeping.ninja",
              "to": "@evedazzle@mas.to",
              "metadata": {
                "handle": "@Pagan_Animist@beekeeping.ninja",
                "body": "\u003cp\u003e\u003cspan class=\"h-card\" translate=\"no\"\u003e\u003ca href=\"https://mas.to/@evedazzle\" class=\"u-url mention\"\u003e@\u003cspan\u003eevedazzle\u003c/span\u003e\u003c/a\u003e\u003c/span\u003e \u003c/p\u003e\u003cp\u003eCan communities band together ...\u003c/p\u003e\u003cp\u003eWould they help?\u003c/p\u003e\u003cp\u003eI’m just thinking of your power grid and next time.\u003c/p\u003e",
                "publication_id": "112840117527501203",
                "timestamp": 1721803551,
                "target": {
                  "handle": "@evedazzle@mas.to",
                  "body": "\u003cp\u003eI have no idea what the other families are doing, or how their hourly workers are coping with having no work for a week and a half. This is a disaster in so many different ways for so many people.\u003c/p\u003e",
                  "publication_id": "112802025232873362",
                  "timestamp": 1721222308
                }
              },
              "related_urls": [
                "https://mas.to/users/evedazzle/statuses/112802025232873362"
              ]
            }
          ],
          "success": true,
          "timestamp": 1721803551
}
```

The `related_urls` are wrong, it should be a link of 112840117527501203.

File: internal/engine/source/activitypub/option.go [hunk 1]
```diff
@@ -18,7 +17,7 @@ type Option struct {
 func NewOption(n network.Network, parameters *config.Parameters) (*Option, error) {
 	var option Option
 
-	fmt.Println("parameters is:", parameters)
+	zap.L().Info("parameters is:", zap.Any("parameters", parameters))
```
Comment (polebug) [inline internal/engine/source/activitypub/option.go:20]: There are some repetitions in the log information, which will form a strange sentence like `parameters is parameters: `. And it should use `are` I think.

File: internal/engine/source/activitypub/task.go [hunk 2]
```diff
@@ -40,7 +40,7 @@ func (t Task) GetTimestamp() uint64 {
 
 	parsedTime, err := time.Parse(time.RFC3339, timeStr)
 	if err != nil {
-		zap.L().Info("Error parsinig time")
+		zap.L().Info("Error parsing time")
```
Comment (polebug) [inline internal/engine/source/activitypub/task.go:43]: It should use `Error`, not `Info`.

File: internal/engine/worker/federated/activitypub/mastodon/worker_test.go [hunk 3]
```diff
@@ -229,29 +232,35 @@ func TestWorker(t *testing.T) {
...
 							Target: &metadata.SocialPost{
-								PublicationID: "https://mas.to/users/evedazzle/statuses/112802025232873362",
+								Handle:        "@evedazzle@mas.to",
+								PublicationID: "112802025232873362",
+								Timestamp:     1721222308,
+								Body:          "<p>I have no idea what the other families are doing, or how their hourly workers are coping with having no work for a week and a half. This is a disaster in so many different ways for so many people.</p>",
 							},
 							Timestamp: 1721803551,
 						},
+						RelatedURLs: []string{"https://mas.to/users/evedazzle/statuses/112802025232873362"},
```
Comment (polebug) [inline internal/engine/worker/federated/activitypub/mastodon/worker_test.go:263]: Url is wrong, it should link of `112840117527501203`.

File: internal/engine/worker/federated/activitypub/mastodon/worker_test.go [hunk 4]
```diff
@@ -296,161 +305,93 @@ func TestWorker(t *testing.T) {
...
-							Target: &metadata.SocialPost{
-								PublicationID: "https://mastodon.social/users/SpookieRobieTheCat/statuses/112840076342641439",
+								Handle:        "@SpookieRobieTheCat@mastodon.social",
+								Body:          "<p><span class=\"h-card\" translate=\"no\"><a href=\"https://mastodon.social/@eunews\" class=\"u-url mention\">@<span>eunews</span></a></span> They need to toss Hungary out.</p>",
+								PublicationID: "112840076342641439",
+								Timestamp:     1721802922,
 							},
 							Timestamp: 1721803252,
 						},
+						RelatedURLs: []string{"https://mastodon.social/users/SpookieRobieTheCat/statuses/112840076342641439"},
```
Comment (polebug) [inline internal/engine/worker/federated/activitypub/mastodon/worker_test.go:336]: The same error.

File: internal/engine/worker/federated/activitypub/mastodon/worker_test.go [hunk 5]
```diff
@@ -296,161 +305,93 @@ func TestWorker(t *testing.T) {
...
-								PublicationID: "https://mock.social/notes/note123",
+								PublicationID: "113289464976693734",
+								Handle:        "@mastodonmigration@mastodon.online",
+								Timestamp:     1728660049,
+								Body:          "<p><span class=\"h-card\" translate=\"no\"><a href=\"https://toad.social/@wdlindsy\" class=\"u-url mention\">@<span>wdlindsy</span></a></span> </p><p>There is a lot of &#39;burn the village to the ground to save the village&#39; about these people.</p>",
 							},
-							Timestamp: 1721606400,
+							Timestamp: 1728549240,
 						},
+						RelatedURLs: []string{"https://mastodon.online/users/mastodonmigration/statuses/113289464976693734"},
```
Comment (polebug) [inline internal/engine/worker/federated/activitypub/mastodon/worker_test.go:391]: The same error.

File: provider/activitypub/mastodon/client.go [hunk 6]
```diff
@@ -32,29 +33,50 @@ func (c *client) GetKafkaConsumer() *kgo.Client {
...
-		}
-	)
+// extractKafkaBroker extracts the host:port part from the given endpoint
+func extractKafkaBroker(endpoint string) string {
+	kafkaBroker := endpoint
 
-	// extract the host:port part for setting kafka consumer service
-	trimmedHostPort := strings.TrimPrefix(endpoint, "http://")
+	// Remove the protocol (e.g., "http://") if present
+	if strings.Contains(kafkaBroker, "://") {
```
Comment (polebug) [inline provider/activitypub/mastodon/client.go:41]: Maybe using `url` pkg would be better.

Comment (brucexc) [reply provider/activitypub/mastodon/client.go]: Since we cannot determine how the user connects to Kafka, we directly use the endpoint as the broker here

**Comment by pseudoyu** [comment]
should add openapi doc of `/federated` APIs

**Review by pseudoyu — APPROVED** [review]

**Pull request merged**